lib/much-timeout.rb

```diff
@@ -1,4 +1,43 @@
+require 'thread'
 require "much-timeout/version"
 
 module MuchTimeout
+
+  TimeoutError = Class.new(Interrupt)
+
+  PIPE_SIGNAL = '.'
+
+  def self.timeout(seconds, klass = nil, &block)
+    if seconds.nil?
+      raise ArgumentError, 'please specify a non-nil seconds value'
+    end
+    if !seconds.kind_of?(::Numeric)
+      raise ArgumentError, "please specify a numeric seconds value "\
+                           "(`#{seconds.inspect}` was given)"
+    end
+    exception_klass = klass || TimeoutError
+    reader, writer  = IO.pipe
+    return_val      = nil
+
+    begin
+      block_thread ||= Thread.new do
+        begin
+          return_val = block.call
+        ensure
+          writer.write_nonblock(PIPE_SIGNAL) rescue false
+        end
+      end
+      if !!::IO.select([reader], nil, nil, seconds)
+        block_thread.join
+      else
+        block_thread.raise exception_klass
+        block_thread.join
+      end
+      return_val
```

Comment:
@kellyredding - Nice catch on the return value. I think technically this pattern of using a local var and closure with the thread block is an anti-pattern because you have the possibility of both threads altering the value. This is obviously the most non-case because it's easily evident that both threads aren't modifying it.

Anyway, you might try using that `value` method on threads here (so this would be `block_thread.value`) and then maybe you could get rid of the local variable `return_val` completely. If that doesn't work then I wouldn't worry about changing this.